src/vm_pmu.rs

```diff
@@ -368,4 +375,195 @@ impl VmPmuState {
+                            });
+                        // Fixed counters must use a static mapping pending finalization of event mapping
+                        match hardware_event_type {
+                            PmuHardware::CpuCycles => {
+                                if is_unconfigured && index == PmuHardware::CpuCycles as usize {
+                                    let counter_index = PmuHardware::CpuCycles as u64;
+                                    configure_counter_csrs(
+                                        config_flags,
+                                        event_type,
+                                        event_data,
+                                        counter_index,
+                                    );
+                                    return Ok(counter_index);
+                                } else {
+                                    return Err(SbiError::InvalidParam);
+                                }
+                            }
+                            PmuHardware::Instructions => {
+                                if is_unconfigured && index == PmuHardware::Instructions as usize {
+                                    let counter_index = PmuHardware::Instructions as u64;
+                                    configure_counter_csrs(
+                                        config_flags,
+                                        event_type,
+                                        event_data,
+                                        counter_index,
+                                    );
+                                    return Ok(counter_index);
+                                } else {
+                                    return Err(SbiError::InvalidParam);
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+                    PmuEventType::Firmware(_) => return Err(SbiError::InvalidParam),
+                    _ => {}
+                }
+                // We can only have non-fixed counters at this point
+                let mut bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+                bitmask_iter
+                    .find(|i| matches!(self.counter_state[*i], NotConfigured))
+                    .map_or_else(|| Err(SbiError::InvalidParam), |i| Ok(i as u64))
+            } else {
+                // If skip_match is set, the counter must already be configured
+                let state = self
+                    .counter_state
+                    .get(counter_index as usize)
+                    .ok_or(SbiError::InvalidParam)?;
+                if matches!(state, Started(_) | Configured(_)) {
+                    Ok(counter_index)
+                } else {
+                    Err(SbiError::InvalidParam)?
+                }
+            }?;
+
+            configure_counter_csrs(config_flags, event_type, event_data, index);
```

Comment:
We shouldn't be re-configuring the event type in the `skip_match` case.

Reply:
Seems like OpenSBI permits re-configuration in the skip match case per my reading of the code (see lib/sbi/sbi_pmu.c:695):

`skip_match:
	if (event_type == SBI_PMU_EVENT_TYPE_HW) {
		if (flags & SBI_PMU_CFG_FLAG_CLEAR_VALUE)
			pmu_ctr_write_hw(ctr_idx, 0);`

Reply:
configure_counter_csrs here updates the shpmevent as well which is reconfiguring the event type. OpenSBI only clears the counter value if the SBI_PMU_CFG_FLAG_CLEAR_VALUE is set.

SBI_PMU_CFG_FLAG_SKIP_MATCH - skips the counter matching part only. But implementation should honor the counter value update flags.